sycl/include/sycl/ext/oneapi/backend/level_zero.hpp

```diff
@@ -125,6 +125,71 @@ inline context make_context<backend::ext_oneapi_level_zero>(
       BackendObject.Ownership == ext::oneapi::level_zero::ownership::keep);
 }
 
+namespace detail {
+inline std::vector<device> get_all_devices(sycl::platform p) {
+  return p.get_devices();
+}
+
+inline std::optional<sycl::device> find_matching_descendent_device(
+    sycl::device d,
+    const backend_input_t<backend::ext_oneapi_level_zero, device>
+        &BackendObject) {
+  if (get_native<backend::ext_oneapi_level_zero>(d) == BackendObject)
+    return d;
+  std::vector<info::partition_property> partition_props =
+      d.get_info<info::device::partition_properties>();
+
+  for (auto pp : partition_props) {
+    if (pp == info::partition_property::partition_by_affinity_domain) {
+      auto sub_devices = d.create_sub_devices<
+          info::partition_property::partition_by_affinity_domain>(
+          info::partition_affinity_domain::next_partitionable);
+      for (auto sub_dev : sub_devices) {
+        if (auto maybe_device =
+                find_matching_descendent_device(sub_dev, BackendObject))
+          return maybe_device;
+      }
+    }
+
+    assert(false && "Unexpected partitioning scheme for a Level-Zero device!");
+  }
+
+  return {};
+}
+} // namespace detail
+
+// Specialization of sycl::make_device for Level-Zero backend.
+// Level-Zero backend specification says:
```

Comment:
I couldn't find a quote supporting that. But even then, it would be impossible to implement. CPU device might have max number of compute units in hundreds, it's impossible to create a fixed hierarchy for partitioning by counts as it explodes combinatorically.

Reply:
My concern is with the root devices, not the subdevices.  This is the relevant quote from the core SYCL spec:

> The execution environment for a SYCL application has a fixed number of [root devices](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#root-device) which does not vary as the application executes.

(https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:device-class)

If `make_device` creates a new `device` object, this statement from the core spec is violated because the number of root devices in the execution environment will no longer be fixed.

Reply:
I've update this PR. @gmlueck can you take another look please?